File: Byte&Buy.AngularSPA/src/app/core/dto/order-delivery/order-delivery-details.ts
import { MoneyDto } from "../common/money-dto";
import { DeliveryChannel } from "../delivery/enum/delivery-channel";

export interface OrderDeliveryDetails {
    carrierCode: string;
    deliveryName: string;
    channel: DeliveryChannel;
    price: MoneyDto;
}

export interface CourierDeliveryDetails extends OrderDeliveryDetails {
    type: 'courier';
    street: string;
    houseNumber: string;
    flatNumber?: string | null;
    city: string;
    postalCity: string;
    postalCode: string;
}

export interface PickupPointDeliveryDetails extends OrderDeliveryDetails {
    type: 'pickupPoint';
    pickupPointName: string;
    pickupPointId: string;
    pickupStreet: string;
    pickupCity: string;
    pickupLocalNumber: string;
}

export interface ParcelLockerDeliveryDetails extends OrderDeliveryDetails {
    type: 'parcelLocker';
    parcelLockerId: string;
}

File: Byte&Buy.Desktop/src/ByteBuy.Services/DTO/OrderDelivery/OrderDeliveryDetails.cs
using ByteBuy.Services.DTO.Delivery.Enum;
using ByteBuy.Services.DTO.Money;
using System.Text.Json.Serialization;

namespace ByteBuy.Services.DTO.OrderDelivery;

public sealed record CourierDeliveryDetails(
    string Street,
    string HouseNumber,
    string? FlatNumber,
    string City,
    string PostalCity,
    string PostalCode) : OrderDeliveryDetails;

public sealed record PickupPointDeliveryDetails(
    string PickupPointName,
    string PickupPointId,
    string PickupStreet,
    string PickupCity,
    string PickupLocalNumber) : OrderDeliveryDetails;

public sealed record ParcelLockerDeliveryDetails(
    string ParcelLockerId) : OrderDeliveryDetails;


[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(PickupPointDeliveryDetails), "pickupPoint")]
[JsonDerivedType(typeof(ParcelLockerDeliveryDetails), "parcelLocker")]
[JsonDerivedType(typeof(CourierDeliveryDetails), "courier")]
public abstract record OrderDeliveryDetails
{
    public string CarrierCode { get; init; } = null!;
    public string DeliveryName { get; init; } = null!;
    public DeliveryChannel Channel { get; init; }
    public MoneyDto Price { get; init; } = null!;
}
